alternator/executor.cc

```diff
@@ -1716,8 +1752,7 @@ class put_item_operation : public rmw_operation {
                _returnvalues == returnvalues::ALL_OLD;
     }
     virtual std::optional<mutation> apply(std::unique_ptr<rjson::value> previous_item, api::timestamp_type ts) const override {
-        if (!verify_expected(_request, previous_item.get()) ||
-            !verify_condition_expression(_condition_expression, previous_item.get())) {
+        if (!verify_conditions(previous_item.get(), _condition_expression)) {
```

Comment:
nitpick: please pick between verify_conditions() being a static function that takes all the parameters (like verify_expected() at al. were) or a member function that has access to all fields. Here you have this function take some fields (like _requests) from members, but the _condition_expression one as a parameter, which is weird.

I personally prefer the "static" approach because it makes it clearer exactly what each function does and needs.

Reply:
It makes little sense to detach verify_conditions from its class by making it static, too many function parameters makes it less readable, class is for encapsulating logic including knowledge of what input parameters are, especially in this case where rmw_operation is encapsulating operation.

It's using _condition_expression as a parameter because every subclass of rmw_operation has a different field (with the same name). Maybe this should be fixed instead by moving the field to rmw_operation?

Reply:
Ok, I didn't remember/realize that _condition_expression is different in that it's not inherited but each has a separate variable with the same name.

I guess the OO vs static function thing is a matter of style, so feel free to keep your OO style. I personally dislike seeing related code get split to different header files and source files - a "static" function just resides in exactly the source file that needs it. I'm not bothered by a static function which takes 5 parameters - quite the opposite - it makes it clear that it needs exactly these 5 parameters, not less and not more.

Reply:
I'm opting for keeping static usage to low, mainly because its widespread in alternator have bitten me a bit when developing expressions cache.

And in general while static functions and dependency injection seem to fulfill Open–closed principle and it's good to know dependencies and functions input in practice when codebase grows overusage leads to code which is very difficult to maintain and modify. As modification is what you need to do, can't really put more layers on layers every time.